R/build_xgboost.R

```diff
@@ -1,58 +1,70 @@
 #' formula version of xgboost
-#' ref: https://www.r-bloggers.com/with-our-powers-combined-xgboost-and-pipelearner/
-data_xgboost <- function(data, x_names, y_name, weight = NULL, nrounds= 10, ...) {
-  # data must be only numeric or logical
-  cols <- quantifiable_cols(data)
-  data <- data[, cols]
-
-  if(any(!x_names %in% colnames(data))){
-    error_cols <- x_names[!x_names %in% colnames(data)]
-    stop(paste0("non numeric column found " + paste0(error_cols, collapse = ", ")))
-  }
-
-  # weight argument should be evaluated using data name space
-  weight <- if (!is.null(substitute(weight))) {
-    lz <- lazyeval::as.lazy(substitute(weight))
-    ret <- lazyeval::lazy_eval(lz, data = data)
-    data <- data[!is.na(ret), ]
-    ret[!is.na(ret)]
-  } else {
-    NULL
-  }
-
-  x <- data.matrix(data[, x_names])
-  if(is.integer(x) || is.logical(x)){
-    # x must be a double matrix
-    x <- matrix(as.double(x), ncol = ncol(x))
-  }
-
-  if (nrow(x) == 0) {
+#' @export
+fml_xgboost <- function(data, formula, nrounds= 10, weights = NULL, watchlist_rate = 0, ...) {
+  term <- terms(formula, data = data)
+  # do.call is used to substitute weight
+  md_frame <- do.call(model.frame, list(term, data = data, weights = substitute(weights)))
+  if(nrow(md_frame) == 0){
     stop("No valid data to create xgboost model.")
+  } else {
+    md_mat <- tryCatch({
+      model.matrix(term, data = md_frame, contrasts = FALSE)
+    }, error = function(e){
+      if(e$message == "contrasts can be applied only to factors with 2 or more levels") {
+        stop("more than 1 unique values are expected for categorical columns assigned as predictors")
+      }
+      stop(e)
+    })
+    weight <- model.weights(md_frame)
+
+    y <- model.response(md_frame)
+
+
+    ret <- if(watchlist_rate != 0.0) {
+      if (watchlist_rate < 0 ||  1 <= watchlist_rate) {
+        stop("watchlist_rate must be between 0 and 1")
+      }
+      # create validation data set
+      index <- sample(seq(nrow(md_mat)), ceiling(nrow(md_mat) * watchlist_rate))
```

Comment:
sample function does randomize
```r
> sample(10, ceiling(10 * 0.3))
[1] 3 9 7
```

xgboost just takes matrix and labels set for validation data, so it has to be created beforehand